fog/ledger/enclave/impl/src/key_image_store.rs

```diff
@@ -0,0 +1,194 @@
+    pub fn new(desired_capacity: u64, logger: Logger) -> Self {
+        Self {
+            omap: Box::new(<ObliviousMapCreator<OSC> as OMapCreator<
+                KeySize,
+                ValueSize,
+                McRng,
+            >>::create(
+                desired_capacity, STASH_SIZE, McRng::default
+            )),
+            logger,
+        }
+    }
+
+    // add a key image containing block index and timestamp
+    pub fn add_record(
+        &mut self,
+        key_image: &KeyImage,
+        block_index: u64,
+        timestamp: u64,
+    ) -> Result<(), AddRecordsError> {
+        let mut value = A8Bytes::<ValueSize>::default();
+        let mut key = A8Bytes::<KeySize>::default(); // key used to add to the oram for key image
+        key.clone_from_slice(&key_image.as_ref());
+        // Flip the first byte of key image, when used as a key in the oblivious
+        // map. This is because we will use key image as a key in the map,
+        // but the map does not support all zeroes as a key. All zeroes is a
+        // valid curve point. But if we flip the first byte, it turns out that
+        // isn't a valid curve point, in the Ristretto group.
+        // So this prevents the OMAP_INVALID_KEY error path.
+        key[0] = !key[0];
+        // write block index data to  value[0..8] write the time stamp data to
+        // value[8..16]
+        value[0..8].clone_from_slice(&block_index.to_le_bytes());
+        value[8..16].clone_from_slice(&timestamp.to_le_bytes());
+        // Note: Passing true means we allow overwrite, which seems fine since
+        // the search_key value is not changing
+        let omap_result_code = self.omap.vartime_write(&key, &value, Choice::from(1));
+        if omap_result_code == OMAP_INVALID_KEY {
+            return Err(AddRecordsError::KeyRejected);
+        } else if omap_result_code == OMAP_OVERFLOW {
+            return Err(AddRecordsError::MapOverflow(
+                self.omap.len(),
+                self.omap.capacity(),
+            ));
+        } else if omap_result_code == OMAP_FOUND {
+            log::debug!(self.logger, "A key image record was clobbered");
+        } else if omap_result_code != OMAP_NOT_FOUND {
+            panic!(
+                "omap_result_code had an unexpected value: {}",
+                omap_result_code
+            );
+        }
+        Ok(())
+    }
+
+    // return new struct KeyImageResult which contains block index and timestamp of
+    // key image as ref to convert key image to 32 bits,
+    // call the oram to query to to key image data
+    pub fn find_record(&mut self, key_image: &KeyImage) -> KeyImageResult {
+        let mut result = KeyImageResult {
+            key_image: *key_image,
+            spent_at: u64::MAX,
+            key_image_result_code: KeyImageResultCode::KeyImageError as u32,
+            timestamp: u64::MAX,
+            timestamp_result_code: 1, /* result code 1 is for the mc watcher is found in the
```

Comment:
please see i added a comment in the code. please let me know if it is good enough

Reply:
I think thats why that crate is no-std...

Reply:
James please see I added the reason in the comment that the timestamp found would not be feasible because it depends on the standard set in the mc watcher. Chris recommended that it was a good reason not to make it a  constant

Reply:
Am I missing something here @tubi70 @garbageslam @jcape?
The `mc-watcher-api` crate is `no_std`: https://github.com/mobilecoinfoundation/mobilecoin/blob/master/watcher/api/src/lib.rs#L5
and we have this constant here: https://github.com/mobilecoinfoundation/mobilecoin/blob/master/watcher/api/src/timestamp.rs#L14

Reply:
@eranrund i guess i was missing something, i assumed that all the `-api` crates like `mc-consensus-api` and `mc-fog-api` and `fog-api` are crates that depend on protobuf and are compatible with grpcio. But that's not actually the case with the watcher api crate.

So the watcher doesn't have any specific grpc API, just the health endpoint? https://github.com/mobilecoinfoundation/mobilecoin/blob/a7e3de17794de6c2ec89c1981ed72c602a890a2e/watcher/src/bin/main.rs#L67
And the API is just, it populates this database, which another process could open?

Reply:
James, Chris and Eran please let me know if you needed me to make a change?

Reply:
I think the thing to do is `use mc_watcher_api::TimestampResultCode`, and make that literal be `TimestampResultCode::TimestampFound as i32` (or whatever type is expected by that struct) instead.

Reply:
yes please try to get this number of `mc-watcher-api` crate at the line that Eran indicated. it seems like it should work. if it doesn't then please report back in this thread.

Reply:
@eranrund is there any kind of test that enforces that the values in that enum match the `watcher.proto` file? if not maybe we should make a ticket for that

Reply:
made the change and now pushing to the branch

Reply:
Chris please let me know if you want to create Ticket for enum match in watcher.proto file